### PR TITLE
[MPDX-8075] Horizontally scroll flows view when dragging near the edges

### PR DESCRIPTION
**Pull request by canac**
## Description

* In the flows view, when dragging near the edges, automatically horizontally scroll the columns to make it easier to drag contacts to off-screen columns on narrow screens

[MPDX-8075](https://jira.cru.org/browse/MPDX-8075)

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)"
- [x] I have applied the appropriate labels. (_Add the label "On Staging" to get the branch automatically merged into staging._)
- [x] I have requested a review from another person on the project

### CODE REVIEW

**Comment by dr-bizz** [comment]
This is amazing! 🤩 How did you build this in the amount of time since you returned? I will review this over the next day or so to ensure everything is working correctly, as Tools are the highest priority.

**Comment by canac** [comment]
> This is amazing! 🤩 How did you build this in the amount of time since you returned? I will review this over the next day or so to ensure everything is working correctly, as Tools are the highest priority.

@dr-bizz Thanks! I thought it was going to be more difficult because I've never implemented something like this before, but it didn't turn out to be too bad, and it seems to work pretty well. No rush, take your time.

**Review by dr-bizz — APPROVED** [review]
I thought I approved this earlier today. My VSC Github must not have submitted the approval. Sorry about that.

This looks great. I tested it on my local and it works. I had a few things, but nothing major.

File: src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.test.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    nextFrame(10);
+    expect(container.scrollBy).toHaveBeenNthCalledWith(2, { left: -3 }); // 300 pixels/s for 10ms = 3px
+    nextFrame(20);
+    expect(container.scrollBy).toHaveBeenNthCalledWith(3, { left: -6 }); // 300 pixels/s for 20ms = 6px
+
+    unmount();
+    expect(cancelAnimationFrame).toHaveBeenCalledWith(4);
+  });
+
+  it('scrolls left when the mouse is near the right edge', () => {
```
[the remark this answers is not in the document]
Comment (canac) [reply src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.test.ts:74]: It's confusing, but we definitely want to `element.scrollBy({ left: 100 })` when the mouse is near the right edge and `element.scrollBy({ left: -100 })` when the mouse is near the left edge. I think it makes the mouse sense to call `scrollBy({ left: 100 })` "scrolling left" and `scrollBy({ left: -100 })` "scrolling right". Scrolling left makes the page shift left, which causes more of the elements at the right edge to be visible.

File: src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    (time) => {
+      // Time since handleScroll was last called in seconds
+      const elapsedTime =
+        lastFrameTime.current === null
+          ? 0
+          : (time - lastFrameTime.current) / 1000;
+      lastFrameTime.current = time;
+
+      if (!enabled || !containerRef.current) {
+        return;
```
[the remark this answers is not in the document]
Comment (canac) [reply src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.ts:41]: Maybe I'm misunderstanding your question, but I don't think so. The return value from this function isn't used anywhere. Returning early will prevent the next animation frame from being scheduled.

File: src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  mouseX: number;
+
+  // How many pixels from the edge should the container start scrolling?
+  scrollThreshold: number;
+
+  // How many pixels per second should the container scroll?
+  scrollVelocity: number;
+}
+
+// This hook automatically scrolls an element horizontally when the mouse nears its edge. It is intended to be
```
[the remark this answers is not in the document]
Comment (canac) [reply src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.ts:20]: Ah good catch! I didn't finish writing this comment lol.

File: src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      // Time since handleScroll was last called in seconds
+      const elapsedTime =
+        lastFrameTime.current === null
+          ? 0
+          : (time - lastFrameTime.current) / 1000;
+      lastFrameTime.current = time;
+
+      if (!enabled || !containerRef.current) {
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (canac) [reply src/components/Contacts/ContactFlow/ContactFlowDragLayer/useAutoScroll.ts]: I had it there before I added the `useEffect` that resets `lastFrameTime`. I think it's OK to move it to the top of the function now.

**Pull request merged**